crates/neon-sys/native/src/neon_threadsafe_cb.h

```diff
@@ -0,0 +1,99 @@
+#ifndef NEON_THREADSAFE_CALLBACK_H_
+#define NEON_THREADSAFE_CALLBACK_H_
+
+#include <uv.h>
+#include "neon.h"
+#include "v8.h"
+#include <mutex>
+
+namespace neon {
+
+    class ThreadSafeCallback {
+    public:
+        ThreadSafeCallback(v8::Isolate *isolate,
+                           v8::Local<v8::Value> self,
+                           v8::Local<v8::Function> callback): isolate_(isolate), close_(false)
+        {
+            async_.data = this;
+            uv_async_init(uv_default_loop(), &async_, async_callback);
+            // Save the this argument and the callback to be invoked.
+            self_.Reset(isolate, self);
+            callback_.Reset(isolate, callback);
+            // Save the context (aka realm) to be used when invoking the callback.
+            context_.Reset(isolate, isolate->GetCurrentContext());
+        }
+
+        void call(void *rust_callback, Neon_ThreadSafeCallbackHandler handler) {
+            {
+                std::lock_guard<std::mutex> lock(mutex_);
+                handlers_.push_back({ rust_callback, handler });
+            }
+            uv_async_send(&async_);
+        }
+
+        void close() {
+            // close is called when the rust struct is dropped
+            // this guarantees that it's called only once and
+            // that no other method (call) will be called after it.
+            close_ = true;
+            uv_async_send(&async_);
+        }
+
+        void async_call() {
+            // Ensure that we have all the proper scopes installed on the C++ stack before
+            // invoking the callback, and use the context (i.e. realm) we saved with the task.
+            v8::Isolate::Scope isolate_scope(isolate_);
+            v8::HandleScope handle_scope(isolate_);
+            v8::Local<v8::Context> context = v8::Local<v8::Context>::New(isolate_, context_);
+            v8::Context::Scope context_scope(context);
+
+            v8::Local<v8::Value> self = v8::Local<v8::Value>::New(isolate_, self_);
+            v8::Local<v8::Function> callback = v8::Local<v8::Function>::New(isolate_, callback_);
+
+            while (true) {
```

Comment:
This loop is the hardest part of the logic to understand. If you could possibly comment it with some explanation, that would help. Also, if you derived this from someone else's code (whether it's the N-API thread-safe callback implementation, or some textbook algorithm), it'd be good to cite the source just so maintainers know where to go in order to understand it better, fix bugs in the future, etc.